igvm/commands.py

```diff
@@ -827,57 +838,60 @@ def _get_hypervisor(hostname, allow_reserved=False):
 def _get_best_hypervisor(vm, hypervisor_states, offline=False):
     hv_env = environ.get('IGVM_MODE', 'production')
 
+    # Get all (theoretically) possible HVs sorted by HV preferences
     hypervisors = (Hypervisor(o) for o in Query({
         'servertype': 'hypervisor',
         'environment': hv_env,
         'vlan_networks': vm.route_network,
         'state': Any(*hypervisor_states),
     }, HYPERVISOR_ATTRIBUTES))
+    hypervisors = sorted_hypervisors(HYPERVISOR_PREFERENCES, vm, hypervisors)
+
+    possible_hvs = OrderedDict()
+    for possible_hv in hypervisors:
+        possible_hvs[str(possible_hv)] = possible_hv
+
+    # Check all HVs in parallel. This will check live data on those HVs
+    # but without locking them. This allows us to do a real quick first
+    # filtering round. Below follows another one on the filtered HVs only.
+    with settings(hosts=possible_hvs.keys(), possible_hvs=possible_hvs):
+        checked_hvs = execute(_check_vm_parallel, vm, offline)
+
+        # Remove unsupported HVs from the list
+        for checked_hv, success in checked_hvs.items():
+            if not success:
+                possible_hvs.pop(checked_hv)
+
+    # No supported HV was found
+    not_found_err = IGVMError(
+        'Cannot find hypervisor matching environment: {}, '
+        'states: {}, vlan_network: {}, offline: {}'.format(
+            hv_env, ', '.join(hypervisor_states), vm.route_network, offline,
+        )
+    )
 
-    for hypervisor in sorted_hypervisors(
-        HYPERVISOR_PREFERENCES, vm, hypervisors
-    ):
-        # The actual resources are not checked during sorting for performance.
-        # We need to validate the hypervisor using the actual values before
-        # the final decision.
-        try:
-            hypervisor.acquire_lock()
-        except InvalidStateError as error:
-            log.warning(error)
-            continue
+    if len(possible_hvs) == 0:
+        raise not_found_err
 
+    # Do another checking iteration, this time with HV locking
+    for possible_hv in possible_hvs.values():
         try:
-            hypervisor.check_vm(vm, offline)
-        except libvirtError as error:
-            hypervisor.release_lock()
-            log.warning(
-                'Preferred hypervisor "{}" is skipped: {}'.format(
-                    hypervisor, error)
-            )
+            possible_hv.acquire_lock()
```

Comment:
Will this lock be release if an exception is thrown ?

Reply:
good catch. fill fix

Reply:
actually when this throws an exception there is in fact no lock. so nothing to unlock